### PR TITLE
Support cffi.FFI.from_buffer

### PR DESCRIPTION
**Pull request by gmarkall**
This can be used to pass contiguous Numpy arrays of float32 and
float64 to C function parameters of types float* and double*
respectively.

Some discussion/notes in #1443.

### CODE REVIEW

File: numba/targets/cffiimpl.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+"""
+Implementation of some CFFI functions
+"""
+
+from __future__ import print_function, absolute_import, division
+
+import cffi
```
Comment (pitrou) [inline numba/targets/cffiimpl.py]: This should fail gracefully if cffi isn't installed.

File: numba/typing/cffi_utils.py [hunk 2]
```diff
@@ -127,6 +141,30 @@ def __init__(self, symbol, cstring):
...
+
+@registry.register
+class FFI_from_buffer(templates.AbstractTemplate):
+    key = cffi.FFI.from_buffer
+
+    def generic(self, args, kws):
+        if kws or (len(args) != 1):
+            return
+        [ary] = args
+        if not (isinstance(ary, types.Array) or ary.layout in ('C', 'F')):
```
Comment (pitrou) [inline numba/typing/cffi_utils.py]: You mean "and", not "or", I presume.

Comment (gmarkall) [inline numba/typing/cffi_utils.py]: Whoops, yes, I do.

**Comment by gmarkall** [comment]
Marking this as WIP because the test failures are indicative of a larger issue, which is that I've typed `cffi.FFI` instances as modules, which is misleading at best, but I consider it a mistake.

**Comment by gmarkall** [comment]
I've fixed up the issues that I'm aware of (test fails, FFI module vs instance, ungraceful failure when CFFI is not installed, and the and/or error checking the array type).

**Comment by seibert** [comment]
This looks great!  Very excited to see this become easier.

**Pull request merged**